### PR TITLE
[MPS] Fix batchnorm for mixed types

### LINKED ISSUE #96113: [mps] [PyTorch 2.0] LayerNorm crashes when input is in `float16`
### 🐛 Describe the bug

As stated in the title, the following crashes when using the `mps` device:

```python
ln = nn.LayerNorm((768,), elementwise_affine=True).to("mps")
ln(torch.randn(1, 77, 768).to("mps", dtype=torch.float16))
```

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230306
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-13.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] functorch==0.2.0a0+b2c9d60
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.2
[pip3] torch==2.1.0.dev20230306
[pip3] torchinfo==1.7.0
[conda] numpy                     1.23.2                   pypi_0    pypi
[conda] torch                     2.1.0.dev20230306          pypi_0    pypi
[conda] torchinfo                 1.7.0                    pypi_0    pypi

cc @ezyang @gchanan @zou3519 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by malfet**
By up/down casting weights to input types

Extend unittests to support float16 input

Fixes https://github.com/pytorch/pytorch/issues/96113

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Normalization.mm [hunk 1]
```diff
@@ -234,7 +234,7 @@ Check if running mean exists (maybe do this check before making graph)
                 MPSGraphTensor* scaledBatchMean = [mpsGraph multiplicationWithPrimaryTensor:batchMeanTensor
                                                                             secondaryTensor:momentumTensor
                                                                                        name:nil];
-                MPSGraphTensor* scaledRunningMean = [mpsGraph multiplicationWithPrimaryTensor:runningMeanTensor
+                MPSGraphTensor* scaledRunningMean = [mpsGraph multiplicationWithPrimaryTensor:castMPSTensor(mpsGraph, runningMeanTensor, input_mps_dtype)
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Normalization.mm:237]: All these casts shouldn't be necessary. If the initial Ranked placeholder is created with correct data type. We shouldn't be adding any casts here in the code.

File: aten/src/ATen/native/mps/operations/Normalization.mm [hunk 2]
```diff
@@ -282,13 +281,14 @@ Check if running mean exists (maybe do this check before making graph)
...
-          MPSGraphTensor* outputTensor = [mpsGraph normalizationWithTensor:inputTensor
-                                                                meanTensor:saveMeanTensor
-                                                            varianceTensor:varTensor
-                                                               gammaTensor:weightTensor
-                                                                betaTensor:biasTensor
-                                                                   epsilon:(float)epsilon
-                                                                      name:nil];
+          const auto inputTensorType = [inputTensor dataType];
+          MPSGraphTensor* outputTensor = [mpsGraph normalizationWithTensor: inputTensor
+                                                                meanTensor: castMPSTensor(mpsGraph, saveMeanTensor, inputTensorType)
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Normalization.mm:286]: The normalization shouldn't need casts here. The computation graph for normalization should be performed in the type  which is requested by the user. Even if it's a pass through, it can add spurious casts which should have been fixed earlier in the Graph to add casts at proper places. My main concern is that it may lead to actual casts in future and will be hidden from the next person modifying the code. I am curious do you need the casts here  for this crash ?

**Review by kulinseth — CHANGES_REQUESTED** [review]
